In [11]:
import numpy as np
import pandas as pd

In [12]:
#Датасет UCI_CBM_DATASET (compressor)
#загружаем датасет,указываем разделитель 'пробел', указываем отсутствие заголовка
data = pd.read_csv(r'UCI CBM Dataset\data.txt', delim_whitespace=True, header=None)

print(data)
#смотрим данные

          0     1          2         3         4        5        6         7   \
0      1.138   3.0    289.964  1349.489  6677.380    7.584    7.584   464.006   
1      2.088   6.0   6960.180  1376.166  6828.469   28.204   28.204   635.401   
2      3.144   9.0   8379.229  1386.757  7111.811   60.358   60.358   606.002   
3      4.161  12.0  14724.395  1547.465  7792.630  113.774  113.774   661.471   
4      5.140  15.0  21636.432  1924.313  8494.777  175.306  175.306   731.494   
...      ...   ...        ...       ...       ...      ...      ...       ...   
11929  5.140  15.0  21624.934  1924.342  8470.013  175.239  175.239   681.658   
11930  6.175  18.0  29763.213  2306.745  8800.352  245.954  245.954   747.405   
11931  7.148  21.0  39003.867  2678.052  9120.889  332.389  332.389   796.457   
11932  8.206  24.0  50992.579  3087.434  9300.274  438.024  438.024   892.945   
11933  9.300  27.0  72775.130  3560.400  9742.950  644.880  644.880  1038.411   

          8        9      1

In [13]:
data.info()
# смотрим информацию о датасете
# все столбцы имеют числовой тип и не содержат пропусков

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11934 entries, 0 to 11933
Data columns (total 18 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       11934 non-null  float64
 1   1       11934 non-null  float64
 2   2       11934 non-null  float64
 3   3       11934 non-null  float64
 4   4       11934 non-null  float64
 5   5       11934 non-null  float64
 6   6       11934 non-null  float64
 7   7       11934 non-null  float64
 8   8       11934 non-null  float64
 9   9       11934 non-null  float64
 10  10      11934 non-null  float64
 11  11      11934 non-null  float64
 12  12      11934 non-null  float64
 13  13      11934 non-null  float64
 14  14      11934 non-null  float64
 15  15      11934 non-null  float64
 16  16      11934 non-null  float64
 17  17      11934 non-null  float64
dtypes: float64(18)
memory usage: 1.6 MB


In [14]:
data.describe()
# смотрим статистические данные о столбцах
# в столбцах 8 и 11 в каждой строке находятся одинаковые значения,
# столбцы 5 и 6 совпадают

0             1             2             3             4   \
count  11934.000000  11934.000000  11934.000000  11934.000000  11934.000000   
mean       5.166667     15.000000  27247.498685   2136.289256   8200.947312   
std        2.626388      7.746291  22148.613155    774.083881   1091.315507   
min        1.138000      3.000000    253.547000   1307.675000   6589.002000   
25%        3.144000      9.000000   8375.883750   1386.758000   7058.324000   
50%        5.140000     15.000000  21630.659000   1924.326000   8482.081500   
75%        7.148000     21.000000  39001.426750   2678.079000   9132.606000   
max        9.300000     27.000000  72784.872000   3560.741000   9797.103000   

                 5             6             7        8             9   \
count  11934.000000  11934.000000  11934.000000  11934.0  11934.000000   
mean     227.335768    227.335768    735.495446    288.0    646.215331   
std      200.495889    200.495889    173.680552      0.0     72.675882   
min        5.304000      5.304000    442.364000    288.0    540.442000   
25%       60.317000     60.317000    589.872750    288.0    578.092250   
50%      175.268000    175.268000    706.038000    288.0    637.141500   
75%      332.364750    332.364750    834.066250    288.0    693.924500   
max      645.249000    645.249000   1115.797000    288.0    789.094000   

                 10            11            12            13            14  \
count  11934.000000  1.193400e+04  11934.000000  11934.000000  11934.000000   
mean       2.352963  9.980000e-01     12.297123      1.029474     33.641261   
std        1.084770  2.533635e-13      5.337448      0.010390     25.841363   
min        1.093000  9.980000e-01      5.828000      1.019000      0.000000   
25%        1.389000  9.980000e-01      7.447250      1.020000     13.677500   
50%        2.083000  9.980000e-01     11.092000      1.026000     25.276500   
75%        2.981000  9.980000e-01     15.658000      1.036000     44.552500   
max        4.560000  9.980000e-01     23.140000      1.052000     92.556000   

                 15           16          17  
count  11934.000000  11934.00000  11934.0000  
mean       0.662440      0.97500      0.9875  
std        0.507132      0.01472      0.0075  
min        0.068000      0.95000      0.9750  
25%        0.246000      0.96200      0.9810  
50%        0.496000      0.97500      0.9875  
75%        0.882000      0.98800      0.9940  
max        1.832000      1.00000      1.0000

In [15]:
data.corr()

0             1         2         3         4         5   \
0   1.000000e+00  9.999152e-01  0.961025  0.962090  0.986017  0.959241   
1   9.999152e-01  1.000000e+00  0.958243  0.960412  0.986621  0.956371   
2   9.610254e-01  9.582428e-01  1.000000  0.989723  0.932993  0.999177   
3   9.620898e-01  9.604117e-01  0.989723  1.000000  0.943038  0.988625   
4   9.860173e-01  9.866210e-01  0.932993  0.943038  1.000000  0.929593   
5   9.592410e-01  9.563708e-01  0.999177  0.988625  0.929593  1.000000   
6   9.592410e-01  9.563708e-01  0.999177  0.988625  0.929593  1.000000   
7   9.612418e-01  9.588406e-01  0.991130  0.979638  0.939762  0.985962   
8            NaN           NaN       NaN       NaN       NaN       NaN   
9   9.827476e-01  9.811609e-01  0.990227  0.989278  0.966659  0.987369   
10  9.630581e-01  9.606126e-01  0.998900  0.995103  0.937460  0.998012   
11           NaN           NaN       NaN       NaN       NaN       NaN   
12  9.691129e-01  9.669746e-01  0.997582  0.996035  0.945938  0.996172   
13  9.533607e-01  9.508128e-01  0.996046  0.994019  0.924231  0.996238   
14  9.136309e-01  9.100139e-01  0.977871  0.962341  0.879120  0.977468   
15  9.314219e-01  9.278150e-01  0.995082  0.980223  0.896959  0.994366   
16  2.264405e-15  6.346298e-16  0.002978  0.001370 -0.018838  0.000754   
17  8.566427e-17  9.367920e-18  0.000358 -0.000018  0.010000  0.000104   

          6         7   8         9         10  11        12        13  \
0   0.959241  0.961242 NaN  0.982748  0.963058 NaN  0.969113  0.953361   
1   0.956371  0.958841 NaN  0.981161  0.960613 NaN  0.966975  0.950813   
2   0.999177  0.991130 NaN  0.990227  0.998900 NaN  0.997582  0.996046   
3   0.988625  0.979638 NaN  0.989278  0.995103 NaN  0.996035  0.994019   
4   0.929593  0.939762 NaN  0.966659  0.937460 NaN  0.945938  0.924231   
5   1.000000  0.985962 NaN  0.987369  0.998012 NaN  0.996172  0.996238   
6   1.000000  0.985962 NaN  0.987369  0.998012 NaN  0.996172  0.996238   
7   0.985962  1.000000 NaN  0.992308  0.989383 NaN  0.990463  0.980071   
8        NaN       NaN NaN       NaN       NaN NaN       NaN       NaN   
9   0.987369  0.992308 NaN  1.000000  0.991704 NaN  0.994373  0.983461   
10  0.998012  0.989383 NaN  0.991704  1.000000 NaN  0.999401  0.997941   
11       NaN       NaN NaN       NaN       NaN NaN       NaN       NaN   
12  0.996172  0.990463 NaN  0.994373  0.999401 NaN  1.000000  0.996339   
13  0.996238  0.980071 NaN  0.983461  0.997941 NaN  0.996339  1.000000   
14  0.977468  0.969655 NaN  0.958711  0.975704 NaN  0.972109  0.974193   
15  0.994366  0.986318 NaN  0.976453  0.992728 NaN  0.989260  0.990981   
16  0.000754 -0.039625 NaN -0.047177  0.008168 NaN  0.008328  0.035285   
17  0.000104 -0.038464 NaN -0.016855 -0.002718 NaN -0.018303  0.011794   

          14        15            16            17  
0   0.913631  0.931422  2.264405e-15  8.566427e-17  
1   0.910014  0.927815  6.346298e-16  9.367920e-18  
2   0.977871  0.995082  2.978258e-03  3.576383e-04  
3   0.962341  0.980223  1.369908e-03 -1.776556e-05  
4   0.879120  0.896959 -1.883800e-02  1.000011e-02  
5   0.977468  0.994366  7.535000e-04  1.044235e-04  
6   0.977468  0.994366  7.535000e-04  1.044235e-04  
7   0.969655  0.986318 -3.962512e-02 -3.846390e-02  
8        NaN       NaN           NaN           NaN  
9   0.958711  0.976453 -4.717657e-02 -1.685505e-02  
10  0.975704  0.992728  8.167586e-03 -2.718186e-03  
11       NaN       NaN           NaN           NaN  
12  0.972109  0.989260  8.327871e-03 -1.830336e-02  
13  0.974193  0.990981  3.528524e-02  1.179423e-02  
14  1.000000  0.985544 -3.203662e-02 -1.887184e-02  
15  0.985544  1.000000 -1.366765e-02 -1.732675e-02  
16 -0.032037 -0.013668  1.000000e+00  4.064242e-14  
17 -0.018872 -0.017327  4.064242e-14  1.000000e+00

In [16]:
data.drop([6,8,11,17], axis=1, inplace=True)
# удаляем столбец 17, т.к. он связан с целевым признаком
# cтолбцы 8 и 11 удаляем, т.к. в них в каждой строке находятся одинаковые значения, они не будут оказывать влияние на итоговый результат
# cтолбец 6 удаляем, т.к. он линейно зависим со столбцом 5, их коэффициент корреляции равен 1

In [17]:
y=data.pop(16)
#присваиваем y целевой признак, удаляем его из датасета

In [18]:
x=data
#присваиваем датасет x

In [19]:
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split, cross_val_score

In [20]:
X_train, X_test, y_train, y_test = train_test_split(x, y, train_size = 0.7)
# разбиваем датасет на выборки

In [21]:
#строим модель линейной регрессии
lin_reg = LinearRegression()

In [22]:
lin_reg.fit(x,y)
#обучаем

LinearRegression()

In [23]:
print('R2 на обучающем :', lin_reg.score(X_train,y_train), '; на тестовом:', lin_reg.score(X_test,y_test))
print('Cross-validation:', np.mean(cross_val_score(lin_reg, x, y, cv=4)))
#смотрим метрику R2
# модель показала неплохой результат на обучающей и тестовой выборке, но плохой на кросс-валидации, 
# которая позволяет дать более объективную оценку ее точности
#R2 на обучающем : 0.8419665534695984 ; на тестовом: 0.845422993153675
#Cross-validation: -3.3756554142599047

R2 на обучающем : 0.8414856416436725 ; на тестовом: 0.8465629135180461
Cross-validation: -3.3756554142599047


In [24]:
scores = cross_val_score(lin_reg, x, y, cv=4)
print('R^2 scores for each fold:', scores)

R^2 scores for each fold: [-5.32293644 -0.81533729 -1.23759505 -6.12675288]


In [25]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

In [26]:
mm_scaler = MinMaxScaler((-1,1))
std_scaler = StandardScaler()

In [27]:
#стандартизируем данные, т.к. значения в разных столбцах имеют сильно отличающиеся порядки
X_mm_scaled = mm_scaler.fit_transform(x)
X_std_scaled = std_scaler.fit_transform(x)

In [28]:
print('Cross-validation:', np.mean(cross_val_score(lin_reg, X_mm_scaled, y, cv=4)))

Cross-validation: -3.375655414381872


In [29]:
print('Cross-validation:', np.mean(cross_val_score(lin_reg, X_std_scaled, y, cv=4)))

Cross-validation: -3.375655414381537


In [30]:
#стандартизация не дала улучшения результатов на кросс-валидации
#Cross-validation: -3.375655414381872
#Cross-validation: -3.375655414381537
#добавим полиномиальных признаков
pol_lin = Pipeline([('poly', PolynomialFeatures(degree=2)),
                    ('regr', LinearRegression())])

In [31]:
print('Linear Cross-validation:', np.mean(cross_val_score(pol_lin, x, y, cv=4)))

Linear Cross-validation: 0.999342045973333


In [32]:
print('Linear Cross-validation:', np.mean(cross_val_score(pol_lin,X_std_scaled , y, cv=4)))

Linear Cross-validation: 0.9995315255906575


In [33]:
print('Linear Cross-validation:', np.mean(cross_val_score(pol_lin, X_mm_scaled, y, cv=4)))

Linear Cross-validation: 0.9995315255916939


результат улучшился
Linear Cross-validation: 0.9995315255906575
Linear Cross-validation: 0.9995315255916939
Применим регуляризацию, для избежания переобучения и борьбы с мультиколлинеарностью

In [34]:
from sklearn.linear_model import Lasso, Ridge
las = Lasso()
rid = Ridge()

las.fit(X_train, y_train)
rid.fit(X_train, y_train)

print('Lasso R2 на обучающем :', las.score(X_train,y_train), '; на тестовом:', las.score(X_test,y_test))
print('Lasso Cross-validation:', np.mean(cross_val_score(las, x, y, cv=4)))

print('Ridge R2 на обучающем :', rid.score(X_train,y_train), '; на тестовом:', rid.score(X_test,y_test))
print('Ridge Cross-validation:', np.mean(cross_val_score(rid, x, y, cv=4)))
#Для модели линейной регриссии без полиномиальных признаков применение регуляризации не дает улучшения результатов

Lasso R2 на обучающем : 0.0 ; на тестовом: -0.00012750119962490203
Lasso Cross-validation: -26.60533276197
Ridge R2 на обучающем : 0.8223849666644593 ; на тестовом: 0.8223799371575544
Ridge Cross-validation: -3.6073133789205443


In [35]:
#прменим регуляризацию и полиномиальные признаки

pol_las = Pipeline([('poly', PolynomialFeatures(degree=2)),
                    ('regr', Lasso(alpha=0.0000001, max_iter=1000000))])

pol_rid = Pipeline([('poly', PolynomialFeatures(degree=3)),
                   ('regr', Ridge(alpha=0.1, max_iter=1000))])
# для модели с регуляризацией типа L1(lasso) увеличино количество итераций, и уменьшен коэффициент alpha для сходимости алшоритма,
# но при данных значениях все еще было предупреждение о несходимости алгоритма, хотя результат на кросс-валидации улучшился
# по сравнению с результатом применения регуляризации типа L1 к линейной модели без добавления полиномиальных признаков
# в сравнении с моделью, где использовалась регуляризацией типа L1(lasso), но не использовались полиномиальные признаки
# для модели с регуляризацией типа L2(ridge) увеличина степень до 3 для решения проблемы с предупреждением о плохой обусловленности матрицы

In [100]:
print('Lasso Cross-validation:', np.mean(cross_val_score(pol_las, x, y, cv=4)))

C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.188e-03, tolerance: 1.091e-04
  model = cd_fast.enet_coordinate_descent(
C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.662e-03, tolerance: 2.383e-04
  model = cd_fast.enet_coordinate_descent(
C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 2.615e-03, toleranc

Lasso Cross-validation: 0.9303057641219858


C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_coordinate_descent.py:628: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 1.953e-03, tolerance: 1.092e-04
  model = cd_fast.enet_coordinate_descent(


In [36]:
print('Ridge Cross-validation:', np.mean(cross_val_score(pol_rid, x, y, cv=4)))
# проверка модели с L2 регуляризацией показала хороший результат на кросс-валидации
# Ridge Cross-validation: 0.9998366990367242

Ridge Cross-validation: 0.9998366990367242


In [37]:
# проверим работу моделей с полиномиальными признаками и регуляризацией для стандартизованных данных
pol_las_std = Pipeline([('poly', PolynomialFeatures(degree=2)),
                    ('scaler', StandardScaler()),
                    ('regr', Lasso(alpha=0.1, max_iter=1000))])
pol_rid_std = Pipeline([('poly', PolynomialFeatures(degree=2)),
                    ('scaler', StandardScaler()),
                    ('regr', Ridge(alpha=0.1, max_iter=1000))])

pol_las_mm = Pipeline([('poly', PolynomialFeatures(degree=2)),
                    ('scaler', MinMaxScaler((-1,1))),
                    ('regr', Lasso(alpha=0.1, max_iter=1000))])
pol_rid_mm = Pipeline([('poly', PolynomialFeatures(degree=2)),
                    ('scaler', MinMaxScaler((-1,1))),
                    ('regr', Ridge(alpha=0.1, max_iter=1000))])

In [38]:
print('Lasso Cross-validation mm:', np.mean(cross_val_score(pol_las_std, X_mm_scaled, y, cv=4)))

Lasso Cross-validation mm: -26.605333999008934


In [39]:
print('Ridge Cross-validation mm:', np.mean(cross_val_score(pol_rid_std, X_mm_scaled, y, cv=4)))

Ridge Cross-validation mm: 0.9391659524121947


In [40]:
print('Lasso Cross-validation std:', np.mean(cross_val_score(pol_las_mm, X_std_scaled, y, cv=4)))

Lasso Cross-validation std: -26.605333999008934


In [41]:
print('Ridge Cross-validation std:', np.mean(cross_val_score(pol_rid_mm, X_std_scaled, y, cv=4)))

Ridge Cross-validation std: 0.9036828732409331


In [42]:
from sklearn.model_selection import GridSearchCV

In [43]:
#Попробуем подобрать для моделей оптимальные параметры коэффициента регуляризации и степени полинома
param_grid = {'poly__degree': range(1, 5),
              'regr__alpha': [0.001, 0.01, 0.1, 5, 10]}

In [44]:
grid_las_std = GridSearchCV(pol_las_std, param_grid, cv=4)

In [45]:
grid_las_std.fit(X_std_scaled,y)

GridSearchCV(cv=4,
             estimator=Pipeline(steps=[('poly', PolynomialFeatures()),
                                       ('scaler', StandardScaler()),
                                       ('regr', Lasso(alpha=0.1))]),
             param_grid={'poly__degree': range(1, 5),
                         'regr__alpha': [0.001, 0.01, 0.1, 5, 10]})

In [46]:
grid_las_std.best_score_

-26.431905299826205

In [47]:
grid_las_std.best_params_

{'poly__degree': 4, 'regr__alpha': 0.001}

In [48]:
grid_las_mm = GridSearchCV(pol_las_mm, param_grid, cv=4)

In [49]:
grid_las_mm.fit(X_mm_scaled,y)

GridSearchCV(cv=4,
             estimator=Pipeline(steps=[('poly', PolynomialFeatures()),
                                       ('scaler',
                                        MinMaxScaler(feature_range=(-1, 1))),
                                       ('regr', Lasso(alpha=0.1))]),
             param_grid={'poly__degree': range(1, 5),
                         'regr__alpha': [0.001, 0.01, 0.1, 5, 10]})

In [50]:
grid_las_mm.best_score_

-26.605333999008934

In [51]:
grid_las_mm.best_params_

{'poly__degree': 1, 'regr__alpha': 0.001}

In [60]:
grid_las = GridSearchCV(pol_las, param_grid, cv=4)

In [52]:
grid_rid_std = GridSearchCV(pol_rid_std, param_grid, cv=4)

In [53]:
grid_rid_std.fit(X_std_scaled,y)

GridSearchCV(cv=4,
             estimator=Pipeline(steps=[('poly', PolynomialFeatures()),
                                       ('scaler', StandardScaler()),
                                       ('regr',
                                        Ridge(alpha=0.1, max_iter=1000))]),
             param_grid={'poly__degree': range(1, 5),
                         'regr__alpha': [0.001, 0.01, 0.1, 5, 10]})

In [54]:
grid_rid_std.best_score_

0.9983287824164235

In [55]:
grid_rid_std.best_params_

{'poly__degree': 4, 'regr__alpha': 0.001}

In [56]:
grid_rid_mm = GridSearchCV(pol_rid_mm, param_grid, cv=4)

In [57]:
grid_rid_mm.fit(X_mm_scaled,y)

GridSearchCV(cv=4,
             estimator=Pipeline(steps=[('poly', PolynomialFeatures()),
                                       ('scaler',
                                        MinMaxScaler(feature_range=(-1, 1))),
                                       ('regr',
                                        Ridge(alpha=0.1, max_iter=1000))]),
             param_grid={'poly__degree': range(1, 5),
                         'regr__alpha': [0.001, 0.01, 0.1, 5, 10]})

In [58]:
grid_rid_mm.best_score_

0.9980191719856956

In [59]:
grid_rid_mm.best_params_

{'poly__degree': 4, 'regr__alpha': 0.001}

In [62]:
grid_rid = GridSearchCV(pol_rid, param_grid, cv=4)

In [63]:
grid_rid.fit(x,y)

C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.50239e-26): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.53131e-26): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.49554e-26): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_ridge.py:211: LinAlgWarning: Ill-conditioned matrix (rcond=2.55077e-26): result may not be accurate.
  return linalg.solve(A, Xy, assume_a="pos", overwrite_a=True).T
C:\Users\tat0e\anaconda3\Lib\site-packages\sklearn\linear_model\_ridge.py:21

GridSearchCV(cv=4,
             estimator=Pipeline(steps=[('poly', PolynomialFeatures(degree=3)),
                                       ('regr',
                                        Ridge(alpha=0.1, max_iter=1000))]),
             param_grid={'poly__degree': range(1, 5),
                         'regr__alpha': [0.001, 0.01, 0.1, 5, 10]})

In [64]:
grid_rid.best_score_

0.9998710375978237

In [65]:
grid_rid.best_params_

{'poly__degree': 3, 'regr__alpha': 10}

вывод: лучшую точность на кросс-валидации показала модель линейной регрессии с добавлением полиномиальных признаков 3-ей степени, которые увеличивают сложность модели и позволяют ей лучше аппроксимировать данные, и L2 регуляризацией с коэффициентом регуляризации 10, который немного снижает её влияние на итоговый результат, но при этом помогает предотвратить переобучение модели.Т.к. входные признаки достаточно сильно коррелирующие, то использование L2 регуляризации более предпочтительно, чем L1 регуляризация. L1 регуляризация склонна оставить только один из входных признаков, занулив веса других признаков. Это может привести к потере информации и ухудшению точности модели. L2 регуляризация стремиться распределить веса равномерно между коррелирующими признаками, что поможет уменьшить их влияние и снизить переобучение.